File: src/SilksongLauncher.Launcher/app/src/main/kotlin/dev/silksong/launcher/skins/core/SkinResourceCleanupModels.kt
package dev.silksong.launcher.skins.core

import java.util.Collections

/** Trusted terminal decoder facts, not stop-request acknowledgement. */
enum class ResourceDecodeOutcome { PENDING, CREATED, NEVER_CREATED }
data class ResourceCleanupDecoder(val operation: ResourceDecodeOperation,
    val outcome: ResourceDecodeOutcome = ResourceDecodeOutcome.PENDING, val scratchReleased: Boolean = false)

/** Exact resource-only shutdown identity, never a fabricated transaction. */
data class ResourceShutdownOperation(val operationId: Long, val plan: ResourceLifetimePlan)
data class ResourceShutdownDisposal(val operationId: Long, val shutdown: ResourceShutdownOperation, val allocationId: Long)
enum class ResourcePostTransactionOutcome { UNKNOWN, VISUAL_ONLY_ROLLBACK, VERIFIED_DURABLE_ROLLBACK }
enum class ResourceRollbackClonePhase { CANDIDATE, OLD_PRIOR, QUALIFIED }
data class ResourceRollbackResolution(val closure: ResourceLifetimeClosure, val restoredPrior: ResourceLifetimePlan?, val clonePhase: ResourceRollbackClonePhase = ResourceRollbackClonePhase.CANDIDATE)

/** This wrapper is the gate authority; never bypass it through its nested lifetime snapshot.
 * All verification events are trusted live adapter facts, not disk proof manufactured here. */
class ResourceCleanupState(val lifetime: ResourceLifetimeState = ResourceLifetimeState(),
    val cancellation: ResourceLifetimeClosure? = null, val decoder: ResourceCleanupDecoder? = null,
    pendingDisposals: Collection<ResourceDisposalOperation> = emptyList(),
    val teardownRequested: Boolean = false, val rollback: ResourceRollbackResolution? = null,
    val shutdown: ResourceShutdownOperation? = null, val shutdownClonesQualified: Boolean = false,
    shutdownDisposals: Collection<ResourceShutdownDisposal> = emptyList()) {
    val pendingDisposals: List<ResourceDisposalOperation> = Collections.unmodifiableList(pendingDisposals.toList())
    val shutdownDisposals: List<ResourceShutdownDisposal> = Collections.unmodifiableList(shutdownDisposals.toList())
    val closed: Boolean get() = teardownRequested && lifetime.nextOperationId > 0 && lifetime.ledger.nextId > 0 &&
        !lifetime.candidateSealed && !lifetime.awaitingDurableCompletion && !shutdownClonesQualified && lifetime.active == null && lifetime.candidate == null &&
        lifetime.awaitingClones == null && lifetime.retirement == null && lifetime.pendingDisposals.isEmpty() &&
        cancellation == null && decoder == null && pendingDisposals.isEmpty() && rollback == null && shutdown == null &&
        shutdownDisposals.isEmpty() && lifetime.ledger.allocations.isEmpty() && lifetime.ledger.references.isEmpty() && lifetime.ledger.scratch == null
}
sealed interface ResourceCleanupEvent {
    data object Teardown : ResourceCleanupEvent
    /** Live adapter verification of durable restored-prior ownership for the exact consumed closure,
     * not visual rollback, command completion, or retained registry history. Restored planId retains
     * the prior ownership group; binding/generation/stamp are supplied, never derived numerically.
     * UNKNOWN/visual-only retain ownership and allow a later exact success or durable rollback event. */
    data class PostTransactionOutcome(val closure: ResourceLifetimeClosure, val outcome: ResourcePostTransactionOutcome,
        val restoredPrior: ResourceLifetimePlan?) : ResourceCleanupEvent
    /** ALL clones of the currently requested exact plan, not just one clone or the restored plan. */
    data class RollbackClonesInvalidated(val closure: ResourceLifetimeClosure, val plan: ResourceLifetimePlan) : ResourceCleanupEvent
    /** Authoritative live registry count for that same full closure and requested plan identity. */
    data class RollbackCloneCountVerified(val closure: ResourceLifetimeClosure, val plan: ResourceLifetimePlan, val count: Long) : ResourceCleanupEvent
    /** ALL current plan clones; an old rollback/retirement ACK cannot qualify this new operation. */
    data class ShutdownClonesInvalidated(val operation: ResourceShutdownOperation) : ResourceCleanupEvent
    /** Authoritative live registry count for the exact shutdown operation and current plan. */
    data class ShutdownCloneCountVerified(val operation: ResourceShutdownOperation, val count: Long) : ResourceCleanupEvent
    data class ShutdownDisposeAcknowledged(val operation: ResourceShutdownDisposal) : ResourceCleanupEvent
    data class Lifetime(val event: ResourceLifetimeEvent) : ResourceCleanupEvent
    data class CancelPreparation(val closure: ResourceLifetimeClosure) : ResourceCleanupEvent
    /** Exact trusted terminal outcome: NEVER_CREATED guarantees no late allocation can be created. */
    data class DecodeResolved(val closure: ResourceLifetimeClosure, val operation: ResourceDecodeOperation,
        val outcome: ResourceDecodeOutcome) : ResourceCleanupEvent
    /** Trusted release of ALL upload/CPU scratch for this operation, independent of outcome arrival. */
    data class ScratchReleased(val closure: ResourceLifetimeClosure, val operation: ResourceDecodeOperation) : ResourceCleanupEvent
    data class DisposeAcknowledged(val operation: ResourceDisposalOperation) : ResourceCleanupEvent
}
sealed interface ResourceCleanupCommand {
    data class InvalidateRollbackClones(val closure: ResourceLifetimeClosure, val plan: ResourceLifetimePlan) : ResourceCleanupCommand
    data class InvalidateShutdownClones(val operation: ResourceShutdownOperation) : ResourceCleanupCommand
    data class DisposeShutdown(val operation: ResourceShutdownDisposal) : ResourceCleanupCommand
    data class Lifetime(val command: ResourceLifetimeCommand) : ResourceCleanupCommand
    data class RequestDecoderStop(val closure: ResourceLifetimeClosure, val operation: ResourceDecodeOperation) : ResourceCleanupCommand
    data class ReleaseUploadScratch(val closure: ResourceLifetimeClosure, val operation: ResourceDecodeOperation) : ResourceCleanupCommand
    data class Dispose(val operation: ResourceDisposalOperation) : ResourceCleanupCommand
}
class ResourceCleanupDecision(val state: ResourceCleanupState, val accepted: Boolean, val diagnosis: String,
    commands: Collection<ResourceCleanupCommand> = emptyList()) {
    val commands: List<ResourceCleanupCommand> = Collections.unmodifiableList(commands.toList())
}

File: tools/hollow-knight-patches/src/skins/core/SkinResourceCleanupModels.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DualSouls.Skins.HollowKnight.Core
{
    // Trusted terminal decoder facts, not stop-request acknowledgement.
    public enum ResourceDecodeOutcome { PENDING, CREATED, NEVER_CREATED }
    public sealed record ResourceCleanupDecoder(ResourceDecodeOperation Operation,
        ResourceDecodeOutcome Outcome = ResourceDecodeOutcome.PENDING, bool ScratchReleased = false);
    // Exact resource-only shutdown identity; never a fabricated transaction.
    public sealed record ResourceShutdownOperation(long OperationId, ResourceLifetimePlan Plan);
    public sealed record ResourceShutdownDisposal(long OperationId, ResourceShutdownOperation Shutdown, long AllocationId);
    public enum ResourcePostTransactionOutcome { UNKNOWN, VISUAL_ONLY_ROLLBACK, VERIFIED_DURABLE_ROLLBACK }
    public enum ResourceRollbackClonePhase { CANDIDATE, OLD_PRIOR, QUALIFIED }
    public sealed record ResourceRollbackResolution(ResourceLifetimeClosure Closure, ResourceLifetimePlan RestoredPrior, ResourceRollbackClonePhase ClonePhase = ResourceRollbackClonePhase.CANDIDATE);
    // This wrapper is the gate authority. Verification events are trusted live adapter facts.
    public sealed class ResourceCleanupState
    {
        public ResourceLifetimeState Lifetime { get; }
        public ResourceLifetimeClosure Cancellation { get; }
        public ResourceCleanupDecoder Decoder { get; }
        public IReadOnlyList<ResourceDisposalOperation> PendingDisposals { get; }
        public bool TeardownRequested { get; }
        public ResourceRollbackResolution Rollback { get; }
        public ResourceShutdownOperation Shutdown { get; }
        public bool ShutdownClonesQualified { get; }
        public IReadOnlyList<ResourceShutdownDisposal> ShutdownDisposals { get; }
        public bool Closed => TeardownRequested && Lifetime.NextOperationId > 0 && Lifetime.Ledger.NextId > 0 &&
            !Lifetime.CandidateSealed && !Lifetime.AwaitingDurableCompletion && !ShutdownClonesQualified && Lifetime.Active == null && Lifetime.Candidate == null &&
            Lifetime.AwaitingClones == null && Lifetime.Retirement == null && Lifetime.PendingDisposals.Count == 0 &&
            Cancellation == null && Decoder == null && PendingDisposals.Count == 0 && Rollback == null && Shutdown == null &&
            ShutdownDisposals.Count == 0 && Lifetime.Ledger.Allocations.Count == 0 && Lifetime.Ledger.References.Count == 0 && Lifetime.Ledger.Scratch == null;
        public ResourceCleanupState(ResourceLifetimeState lifetime = null, ResourceLifetimeClosure cancellation = null,
            ResourceCleanupDecoder decoder = null, IEnumerable<ResourceDisposalOperation> pendingDisposals = null,
            bool teardownRequested = false, ResourceRollbackResolution rollback = null,
            ResourceShutdownOperation shutdown = null, bool shutdownClonesQualified = false,
            IEnumerable<ResourceShutdownDisposal> shutdownDisposals = null)
        {
            Lifetime = lifetime ?? new ResourceLifetimeState(); Cancellation = cancellation; Decoder = decoder;
            PendingDisposals = Array.AsReadOnly((pendingDisposals ?? Array.Empty<ResourceDisposalOperation>()).ToArray());
            TeardownRequested = teardownRequested; Rollback = rollback; Shutdown = shutdown; ShutdownClonesQualified = shutdownClonesQualified;
            ShutdownDisposals = Array.AsReadOnly((shutdownDisposals ?? Array.Empty<ResourceShutdownDisposal>()).ToArray());
        }
    }
    public abstract record ResourceCleanupEvent
    {
        private ResourceCleanupEvent() { }
        public sealed record Teardown : ResourceCleanupEvent;
        // Live adapter verification of durable restored-prior ownership for this exact consumed closure,
        // not visual rollback, command completion, or retained registry history. Restored PlanId retains
        // prior ownership; binding/generation/stamp are supplied, never derived numerically.
        // UNKNOWN/visual-only preserve ownership; later exact success/durable rollback remains reachable.
        public sealed record PostTransactionOutcome(ResourceLifetimeClosure Closure, ResourcePostTransactionOutcome Outcome, ResourceLifetimePlan RestoredPrior) : ResourceCleanupEvent;
        // ALL clones of the requested exact plan, not one clone or the restored plan.
        public sealed record RollbackClonesInvalidated(ResourceLifetimeClosure Closure, ResourceLifetimePlan Plan) : ResourceCleanupEvent;
        // Authoritative live registry count for the same full closure and requested plan identity.
        public sealed record RollbackCloneCountVerified(ResourceLifetimeClosure Closure, ResourceLifetimePlan Plan, long Count) : ResourceCleanupEvent;
        // ALL current plan clones; old rollback/retirement ACK cannot qualify this new operation.
        public sealed record ShutdownClonesInvalidated(ResourceShutdownOperation Operation) : ResourceCleanupEvent;
        // Authoritative live registry count for the exact shutdown operation and current plan.
        public sealed record ShutdownCloneCountVerified(ResourceShutdownOperation Operation, long Count) : ResourceCleanupEvent;
        public sealed record ShutdownDisposeAcknowledged(ResourceShutdownDisposal Operation) : ResourceCleanupEvent;
        public sealed record Lifetime(ResourceLifetimeEvent Event) : ResourceCleanupEvent;
        public sealed record CancelPreparation(ResourceLifetimeClosure Closure) : ResourceCleanupEvent;
        // NEVER_CREATED guarantees no late allocation can be created; trusted terminal adapter fact.
        public sealed record DecodeResolved(ResourceLifetimeClosure Closure, ResourceDecodeOperation Operation, ResourceDecodeOutcome Outcome) : ResourceCleanupEvent;
        // Release of ALL upload/CPU scratch, independently arriving from decoder outcome.
        public sealed record ScratchReleased(ResourceLifetimeClosure Closure, ResourceDecodeOperation Operation) : ResourceCleanupEvent;
        public sealed record DisposeAcknowledged(ResourceDisposalOperation Operation) : ResourceCleanupEvent;
    }
    public abstract record ResourceCleanupCommand
    {
        private ResourceCleanupCommand() { }
        public sealed record InvalidateRollbackClones(ResourceLifetimeClosure Closure, ResourceLifetimePlan Plan) : ResourceCleanupCommand;
        public sealed record InvalidateShutdownClones(ResourceShutdownOperation Operation) : ResourceCleanupCommand;
        public sealed record DisposeShutdown(ResourceShutdownDisposal Operation) : ResourceCleanupCommand;
        public sealed record Lifetime(ResourceLifetimeCommand Command) : ResourceCleanupCommand;
        public sealed record RequestDecoderStop(ResourceLifetimeClosure Closure, ResourceDecodeOperation Operation) : ResourceCleanupCommand;
        public sealed record ReleaseUploadScratch(ResourceLifetimeClosure Closure, ResourceDecodeOperation Operation) : ResourceCleanupCommand;
        public sealed record Dispose(ResourceDisposalOperation Operation) : ResourceCleanupCommand;
    }
    public sealed class ResourceCleanupDecision
    {
        public ResourceCleanupState State { get; }
        public bool Accepted { get; }
        public string Diagnosis { get; }
        public IReadOnlyList<ResourceCleanupCommand> Commands { get; }
        public ResourceCleanupDecision(ResourceCleanupState state, bool accepted, string diagnosis, params ResourceCleanupCommand[] commands)
        { State = state; Accepted = accepted; Diagnosis = diagnosis; Commands = Array.AsReadOnly((ResourceCleanupCommand[])commands.Clone()); }
    }
}
